### DISCUSSION TITLE
Best practice for resetting app state on events like sign out

### DISCUSSION THREAD

**Discussion by aardrop** [rrousselGit/riverpod#2679] [Q&A]
I wonder if people have a best practice for resetting their app on events like signout. I have tried things like Pheonix but always come into the disposed provider called. I want to reset the entire app state back to the initial launch on signout, but I can't seem to do this cleanly. 

Do others have recommendations or a best practice for things like this?

**Comment by rrousselGit** [comment]
You should "ref.watch" the login state inside providers which need to be reset

**Reply by aardrop** [reply]
Apologies allow me to clarify, I have tons of pages and providers and some listen to the login state others don't but all of them have to be completely reset on signout. I've tried refresh and invalidate but when we have over 40 providers it's convoluted

**Reply by JulianBissekkou** [reply]
We have the same problem and we solved it by recreating the `ProviderContainer` and using a `UnconrolledProviderScope`.
However, this is quite a workaround and I was wondering if there is a better solution that is less verbose.

@rrousselGit

**Reply by rrousselGit** [reply]
You aren't supposed to reset everything. That's a big no-no generally, as you would fetch way more than needed.

There are many ways to deal with this. For example, you could create single Repository hosting all the API methods depending on the user. And have that repo ref.watch the token state.

Then if you were to do `ref.watch(userRepositoryProvider))`, all providers using that repository would naturally update.

**Reply by aardrop** [reply]
I understand that.

Generally, if our environment is different for each user, we would want to reload every single thing if the user was to log out and log in under a new account.

So if we aren't supposed to reset everything does that mean we need to use methods like union classes to manual reset the state of each provider?

**Reply by temcewen** [reply]
I think you missed the point of what @rrousselGit was saying. If you're using Firebase for authentication, you can turn your firebase token stream into a Riverpod stream and then watch that in whatever providers should be reset when the user logs out. 

Also, should be noted that if your providers are set to autodispose, they should naturally be reset when all the pages are closed due to the user signing out. I believe that autodisposable providers are preferred as they are intrinsically more efficient, so perhaps you can consider transitioning to using more of those.

**Comment by Hamid313-coder** [comment]
Having the same issue with log out.
any solution?

**Reply by ash1497** [reply]
I am using provider, all my viewmodels are declared in MultiProviders at the entry level (main.dart). What I did was that I used a keyedSubtree and encompassed the MultiProvider in this widget. essentially all I do now is trigger this upon signout and it essentially resets the app. I can provide the code I used:

import 'package:flutter/material.dart';

class RestartWidget extends StatefulWidget {
  final Widget child;

  const RestartWidget({super.key, required this.child});

  static void restartApp(BuildContext context) {
    final _RestartWidgetState? state =
        context.findAncestorStateOfType<_RestartWidgetState>();
    state?.restartApp();
  }

  @override
  // ignore: library_private_types_in_public_api
  _RestartWidgetState createState() => _RestartWidgetState();
}

class _RestartWidgetState extends State<RestartWidget> {
  Key key = UniqueKey();

  void restartApp() {
    setState(() {
      key = UniqueKey();
    });
  }

  @override
  Widget build(BuildContext context) {
    return KeyedSubtree(
      key: key,
      child: widget.child,
    );
  }
}

And then:


runApp(
    RestartWidget(
      child: MultiProvider(
        providers: [
         ALL YOUR PROVIDERS
         ]
    ),
  ),
); 
  
In my signout function, I simply call:
RestartWidget.restartApp(context);
This essentially resets the app as if I fresly launched it, in my case it navigates me right on the sign-in page since that's the entry point if the user isn't signed in. 

This was my solution, Not sure if its the best solution but it works for me in the scenario where the user logsout of one account, then logs back into another account right away, the state is refreshed and no data overlap occurs. 

If someone has a better solution (Since I am not using RiverPod, just Provider) Please correct/add onto this.

**Reply by aardrop** [reply]
It's not perfect, but to solve this, we've had to manually manage the disposal of providers like auth, user, and their dependencies. 

Our app allows unauthenticated users access to the platform before logging in and then has multiple pages, widgets, and features that unlock once authenticated. As a result we can't just dispose of higher-level widgets and cascade that. So we manually manage the disposal, refresh, and invalidating of providers on an event basis, such as sign out, so that we clear the app state in the reverse order it was constructed without breaking any of the providers or needing to rebuild our main widget page. 

I'd love a better way to do this for Riverpod if anyone has a better example, but that is how we handle it.

**Reply by ash1497** [reply]
fair enough, your specific use-case definitely invalidates my approach since you're serving different content based on auth status, that definitely complicates things, but it seems you have figured out a working solution. In my case, we're building a social media app, so everything is served under the auth umbrella, so for me, I suppose what I have here is more of a brute force approach, but I would rather do this than inject an entire library.

**Comment by stefanorumi7** [comment]
I am facing the same issue over here. Can't believe there is no easy way to reset the state to "as if the app was just opened" to ensure we are not leaving any state at all from the previous user and just be safe.

**Reply by rrousselGit** [reply]
That's just a bad solution.
That's like turning off the entire grid system because one person wanted to turn off his light.

**Reply by ash1497** [reply]
Are you referring to what I posted? Can't tell with the way GitHub structures the comments, but if it is, then yeah I definitely don't think I had the best solution, but I couldn't find anything else at that time to work with provider and this is such an infrequent situation, I just wanted to have something in place to deal with it. Either way, this project was over a long time ago, so doesn't matter at this point, but definitely curious what would work better if someone has answers.

**Comment by TDuffinNTU** [comment]
There's event-driven architectures and streams you could use for handling this maybe. 

If your provider Foo _must be reset_ when provider Bar's value swaps from true -> false then you _actually_ want to either:

1. `ref.watch(barProvider), or`
2. `ref.listen(barProvider, (_, next) => next == false ? dispose() : null)`

Both of those do result in you having a _lot_ of coupling between barProvider and others, but this is either not an issue, or you can use some sort of event stream to decouple it. Just provide the stream to everyone and let anyone dump an event onto it.

If the provider isn't disposed on logout, but depends on providers that rebuild due to logout/login, then rebuilds will cascade. Granted, this is complicated and I'm struggling with how to write it myself...
